### ISSUE TITLE
Bug: [no-for-in-array] fails to report on `Array<T> | undefined`

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10534]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gWygTwCMQBBGGALnkxgEsUBzAbQF14AfeAVxVADM6QwANwAoUJFgJk6LLApUMtBi1EiA9Gvi8oNCGkRIy4DBDwjeh%2BAAppmeDQwhs9lPDkBKeAG8AviPWaaAAWSJwQwFo6EAA0VFDYCADuDkHwAAYAKizs8ChhEGnmljaodg5OLjj4RKQwnr7%2BGvAADkiOKBg0UBCmVJz09CCYalCcGEj8AB7w4zNBCBY9SMkMAPxFMNa2WOXOE1WEJGTwq6vwLPV%2BQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oDMPpbmtAIbRoQwuiiJRfSODABfEAqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const maybeArr: string[] | undefined;
declare const arr: string[];

// fails correctly
for (const item in arr) {}

// should fail, same with `T[] | null`
for (const item in maybeArr) {}

// potentially suggest/autofix to the following?
for (const item of maybeArr ?? []) {}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-for-in-array": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the rule to report, as it doesn't make sense to iterate `null` or `undefined`. A fixer would be nice (in general, not just for this special case) since most of the time, the intent is to use `for-of`.

### Actual Result

The rule fails to catch these cases.

### Additional Info

Currently, the rule checks specifically for an array or union of arrays. I think this can be expanded to a union with an array and potentially `null` and `undefined`.

**Comment by kirkwaiblinger** [comment]
I'm sort of wondering whether we should just report any union includes an array, kind of like we do for no-base-to-string, rather than only still reporting when `null`/`undefined` are present. Is there a problem with that approach?

Like, why not also report this?
```ts
declare const x: string[] | Record<number, string>;
for (const k in x) {}
```

**Comment by ronami** [comment]
> I'm sort of wondering whether we should just report any union includes an array, kind of like we do for no-base-to-string, rather than only still reporting when `null`/`undefined` are present. Is there a problem with that approach?
> 
> Like, why not also report this?
> 
> ```ts
> declare const x: string[] | Record<number, string>;
> for (const k in x) {}
> ```

Hmm, that's interesting. I'm +1 to doing this. There are definitely similarities between this and `no-base-to-string`, which would essentially mean a warning on anything that may be an array (possibly with a different message if the value may be an array).

I'll update the open PR to match this 👍

**Comment by kirkwaiblinger** [comment]
Ok! Seeing a positive sentiment from 3 of us, let's plan for that solution, then. IMO, it's definitely not a requirement to special-case the "will"/"may" distinction for this rule, but don't let that stop you from being fancy either if you want 🙂
